File: recursion-tutorial-fc/fibonacci.js
const fib = (n, arr = []) => {

    if(arr[n]) return arr[n];

    if( n <= 0) return 0;
    if( n == 1) return 1;

    arr[n] = fib(n - 1, arr) + fib (n -2, arr);
    return arr[n];
}

console.log(fib(0));
console.log(fib(-1));
console.log(fib(2));
console.log(fib(3));
console.log(fib(4));
console.log(fib(5));
console.log(fib(6));

File: recursion-tutorial-fc/Fibonacci.java
import java.util.Map;
import java.util.HashMap;

public class Fibonacci {
    
    public static Integer fib(int n, Map<Integer, Integer> cache)  {

        if(cache.containsKey(n)) return cache.get(n);
        
        int result = fib(n-2, cache) + fib(n-1, cache);
        cache.put(n, result);

        return result;
    }
    public static void main(String[] args) {

        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 0);
        map.put(1, 1);

        System.out.println(Fibonacci.fib(6, map));
        
    }
}
